### PR TITLE
Fix losing UTC DateTimeKind on ISO8601 UTC values

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/aspnet/SignalR/issues/2305

### CODE REVIEW

**Comment by davidfowl** [comment]
Can we do this as 2 separate fixes? The date fixes should be one and the end of array fixes should be another one. Don’t we also need to handle return types as well as arguments for dates.

File: src/Microsoft.AspNetCore.SignalR.Protocols.Json/Protocol/JsonHubProtocol.cs [hunk 1]
```diff
@@ -599,14 +605,48 @@ private HubMessage BindInvocationMessage(string invocationId, string target, obj
...
+            else if (type == typeof(DateTimeOffset) || type == typeof(DateTimeOffset?))
+            {
+                reader.ReadAsDateTimeOffset();
+            }
+            else
+            {
+                reader.Read();
+            }
+
+            return reader.TokenType != JsonToken.None;
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Protocols.Json/Protocol/JsonHubProtocol.cs]: Comment on what this means.

File: src/Microsoft.AspNetCore.SignalR.Protocols.Json/Protocol/JsonHubProtocol.cs [hunk 2]
```diff
@@ -599,14 +605,48 @@ private HubMessage BindInvocationMessage(string invocationId, string target, obj
...
 
+        private static bool ReadForType(JsonTextReader reader, Type type)
+        {
+            // Explicity read values as dates from JSON with reader.
+            // We do this because otherwise dates are read as strings
+            // and the JsonSerializer will use a conversion method that won't
+            // preserve UTC in DateTime.Kind for UTC ISO8601 dates
+            if (type == typeof(DateTime) || type == typeof(DateTime?))
+            {
+                reader.ReadAsDateTime();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Protocols.Json/Protocol/JsonHubProtocol.cs]: This changes the token type to Date right? Do we have anything else like this?

File: src/Microsoft.AspNetCore.SignalR.Protocols.Json/Protocol/JsonHubProtocol.cs [hunk 3]
```diff
@@ -599,14 +605,48 @@ private HubMessage BindInvocationMessage(string invocationId, string target, obj
             return new InvocationMessage(invocationId, target, arguments);
         }
 
+        private static bool ReadForType(JsonTextReader reader, Type type)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Protocols.Json/Protocol/JsonHubProtocol.cs]: Move this to JsonUtils

File: src/Microsoft.AspNetCore.SignalR.Protocols.Json/Protocol/JsonHubProtocol.cs (hunk 2 above)
Comment (JamesNK) [reply src/Microsoft.AspNetCore.SignalR.Protocols.Json/Protocol/JsonHubProtocol.cs]: Yes.

What?

Comment (davidfowl) [reply src/Microsoft.AspNetCore.SignalR.Protocols.Json/Protocol/JsonHubProtocol.cs]: I was looking at `JsonTokenType` and JsonSerializerSettings to see if there are any other things in this category. The only one I see is FloatParseHandling.

Comment (JamesNK) [reply src/Microsoft.AspNetCore.SignalR.Protocols.Json/Protocol/JsonHubProtocol.cs]: Only date reading was disable in the reader.

**Pull request merged**